executor/analyze.go

```diff
@@ -381,3 +401,444 @@ func (e *AnalyzeColumnsExec) buildStats() (hists []*statistics.Histogram, cms []
+			collectors[hasPKInfo+j].Samples[samplePos].Value, err = tablecodec.DecodeColumnValue(value, &colInfo.FieldType, timeZone)
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+		for j, idxInfo := range e.idxsInfo {
+			collectors[len(e.colsInfo)+hasIdxInfo+j].Samples[samplePos].Ordinal = int(samplePos)
+			idxVals := make([]types.Datum, 0, len(idxInfo.Columns))
+			for _, idxCol := range idxInfo.Columns {
+				val, err := tablecodec.DecodeColumnValue(value, &e.table.Cols()[idxCol.Offset].FieldType, timeZone)
+				if err != nil {
+					return errors.Trace(err)
+				}
+				idxVals = append(idxVals, val)
+			}
+			var bytes []byte
+			bytes, err = codec.EncodeKey(e.ctx.GetSessionVars().StmtCtx, bytes, idxVals...)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			collectors[len(e.colsInfo)+hasIdxInfo+j].Samples[samplePos].Value = types.NewBytesDatum(bytes)
+		}
+		samplePos++
+	}
+	return nil
+}
+
+func (e *AnalyzeFastExec) handleScanIter(iter kv.Iterator, collectors []*statistics.SampleCollector) (err error) {
+	timeZone := e.ctx.GetSessionVars().Location()
+	hasPKInfo := 0
+	if e.pkInfo != nil {
+		hasPKInfo = 1
+	}
+	hasIdxInfo := len(e.idxsInfo)
+	for ; iter.Valid() && err == nil; err = iter.Next() {
+		// reservoir sampling
+		e.rowCount++
+		randNum := rand.Int63n(int64(e.rowCount))
+		if randNum > int64(maxSampleSize) && e.sampCursor == maxSampleSize {
+			continue
+		}
+
+		p := rand.Int63n(int64(maxSampleSize))
+		if e.sampCursor < maxSampleSize {
+			p = int64(e.sampCursor)
+			e.sampCursor++
+		}
+		if hasPKInfo > 0 {
+			collectors[0].Samples[p].Value, err = tablecodec.DecodeColumnValue(iter.Value(), &e.pkInfo.FieldType, timeZone)
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+		for j, colInfo := range e.colsInfo {
+			collectors[hasPKInfo+j].Samples[p].Value, err = tablecodec.DecodeColumnValue(iter.Value(), &colInfo.FieldType, timeZone)
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+		for j, idxInfo := range e.idxsInfo {
+			idxVals := make([]types.Datum, 0, len(idxInfo.Columns))
+			for _, idxCol := range idxInfo.Columns {
+				val, err := tablecodec.DecodeColumnValue(iter.Value(), &e.table.Cols()[idxCol.Offset].FieldType, timeZone)
+				if err != nil {
+					return errors.Trace(err)
+				}
+				idxVals = append(idxVals, val)
+			}
+			var bytes []byte
+			bytes, err = codec.EncodeKey(e.ctx.GetSessionVars().StmtCtx, bytes, idxVals...)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			collectors[len(e.colsInfo)+hasIdxInfo+j].Samples[p].Value = types.NewBytesDatum(bytes)
+		}
+	}
+	return errors.Trace(err)
+}
+
+func (e *AnalyzeFastExec) handleScanTasks(bo *tikv.Backoffer, collectors []*statistics.SampleCollector) error {
+	snapshot, err := e.ctx.GetStore().(tikv.Storage).GetSnapshot(kv.MaxVersion)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	for _, t := range e.scanTasks {
+		iter, err := snapshot.Iter(t.StartKey, append(t.EndKey, '0'))
```

Comment:
Why append `0`? It is exclusive.

Reply:
The iterator iter the range [startKey, endKey).

Reply:
Yes, and the `EndKey` of region is also exclusive.

Reply:
OK